gr-digital/include/gnuradio/digital/lfsr.h

```diff
@@ -90,32 +90,26 @@ namespace digital {
 class lfsr
 {
 private:
-    uint32_t d_shift_register;
-    uint32_t d_mask;
-    uint32_t d_seed;
-    uint32_t d_shift_register_length; // less than 32
-
-    static uint32_t popCount(uint32_t x)
-    {
-        uint32_t r = x - ((x >> 1) & 033333333333) - ((x >> 2) & 011111111111);
-        return ((r + (r >> 3)) & 030707070707) % 63;
-    }
+    uint64_t d_shift_register;
+    uint64_t d_mask;
+    uint64_t d_seed;
+    uint64_t d_shift_register_length; // less than 32
 
 public:
-    lfsr(uint32_t mask, uint32_t seed, uint32_t reg_len)
+    lfsr(uint64_t mask, uint64_t seed, uint64_t reg_len)
         : d_shift_register(seed),
           d_mask(mask),
           d_seed(seed),
           d_shift_register_length(reg_len)
     {
-        if (reg_len > 31)
-            throw std::invalid_argument("reg_len must be <= 31");
+        if (reg_len > 63)
+            throw std::invalid_argument("reg_len must be <= 63");
     }
 
     unsigned char next_bit()
     {
         unsigned char output = d_shift_register & 1;
-        unsigned char newbit = popCount(d_shift_register & d_mask) % 2;
+        unsigned char newbit = __builtin_parityl(d_shift_register & d_mask);
```

Comment:
We have people who compile GR with MSVC. So, this is probably not the way to go. I saw in the discussions that you switched to VOLK for this. That'd be the way to go I assume.

Reply:
Yes, and first I thought I got good results with VOLK popcnt but I cannot reproduce it, I am sorry to say I think I made a mistake and did performance-tests still on the __builtin-code and not VOLK. I only get a minor improvement with VOLK from HEAD. Perhaps like ThomasHabets suggested, this function should be added to VOLK, and use the __builtin_parity when available? If I was better at MSVC, perhaps this could be resolved, because what __builtin does is use a CPU-feature for this calculation. In VOLK, it could be added as inline assemlby if the arch is right?

Reply:
I see these options:

1. Add this to volk either as inline asm or just using `__builtin_xxx` where available, and wait for gnuradio to have a minimum volk version that includes this.
2. Have CMake in gnuradio detect existence of this builtin, and fall back to volk popcount where not available.
3. Use `#ifdef` to check for volk support, then (2), in that order. And also implement in volk.

(2) and (3) can of course also have an MSVC option, but I also don't know what that would be.

Seems the decision could depend on if we think 3.9 can depend on a newer volk, and on how critical performance is in this code.

Reply:
I just spent some time to get into VOLK and when I try to use __builtin_parity from a kernel I implemented there, there is overhead that more than doubles the time than calling __builtin_parity directly, from a little test-program. I _must_ be doing something wrong, but I don't know what.

I like your option 2 the best, there is a simple #ifdef __GNUC__ that I can use for this. Someone with knowledge in MSVC may help to complement on that. Might not look pretty inside a block like that but right know that's my best shot.

Reply:
Does you volk solution end up in a header file as static inline? Otherwise yeah the call would eat up most or all of the gain.

Reply:
This is what I came up with, and when I follow the call to volk in gdb, there are alot of calls before I end up where I want to. Should I try something else?

 67 static inline void volk_64u_parity_generic(int32_t* res, uint64_t value)
 68 {
 69     *res = __builtin_parityl(value);
 70 }

Reply:
VOLK is usually more oriented towards longer vectors. So the overhead may outweigh the benefit here.
Though, benchmarks are notoriously fragile. e.g. did you run a couple of iterations before you started to measure? 
VOLK registers the correct function during runtime. So this might include overhead at first.

Reply:
Is there a way to skip the function registration benchmarking thing, and just "know" that `__builtin_parityl` is the fastest?

As for `#ifdef GNUC`, I would say that it's an *antipattern* to check for which environment you're in. You should check for *capability*. CMake surely has a way of checking, and then give you `ifdef HAVE_BUILTIN_PARITYL`?